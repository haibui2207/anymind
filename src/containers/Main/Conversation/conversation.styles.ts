import { createUseStyles } from "react-jss";

import { rem } from "../../../utils/jss";
import COLORS from "../../../constants/COLORS";

export default createUseStyles({
  root: {
    flex: 1,
    boxSizing: "border-box",
    display: "flex",
    flexDirection: "column",
    "& .ck": {
      "&.ck-editor": { flex: 1 },
      "&.ck-toolbar": { display: "none" },
      "&.ck-editor__main": { height: "100%" },
      "&.ck-editor__editable": {
        height: "100%",
        "& p": { fontSize: rem(14), lineHeight: rem(20) },
      },
      "&.ck-editor__editable_inline": {
        border: "none",
        "&::-webkit-scrollbar": { width: 10, height: 3 },
        "&::-webkit-scrollbar-track, &::-webkit-scrollbar-thumbhover": {
          background: "transparent",
        },
        "&::-webkit-scrollbar-thumb": {
          background: "rgba(178, 178, 178, 0.6)",
          borderRadius: 10,
        },
      },
    },
  },
  header: {
    flex: "none",
    margin: 0,
    padding: rem(16),
    fontSize: rem(24),
    lineHeight: rem(32),
    borderBottom: `1px solid ${COLORS.borderBlue}`,
  },
  content: {
    flex: "1 0 0",
    overflow: "auto",
    borderBottom: `1px solid ${COLORS.borderBlue}`,
  },
  chatBox: {
    flex: "none",
    display: "flex",
    minHeight: rem(40),
    maxHeight: rem(120),
  },
  sendButton: {
    flex: "none",
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    height: "100%",
    padding: rem(12),
    width: rem(80),
    cursor: "pointer",
    border: "none",
    color: COLORS.textWhite,
    backgroundColor: COLORS.bgBlack,
    transition: "opacity .2s ease",
    "&:not(:disabled):hover": { opacity: 0.85 },
    "&:disabled": { opacity: 0.5, cursor: "not-allowed" },
  },
});
