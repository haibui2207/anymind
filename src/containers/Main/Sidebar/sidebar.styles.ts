import { createUseStyles } from "react-jss";

import { rem } from "../../../utils/jss";
import COLORS from "../../../constants/COLORS";
import BREAKPOINTS from "../../../constants/BREAKPOINTS";

export default createUseStyles({
  root: {
    width: rem(320),
    boxSizing: "border-box",
    padding: [rem(16), rem(16), rem(24)],
    backgroundColor: COLORS.bgBlack,
    height: "100%",
    transform: "translateX(0)",
    transition: "width .5s ease, padding .3s ease, transform .5s ease",
  },
  infoWrapper: { display: "flex", alignItems: "center", marginBottom: rem(32) },
  infoTitle: {
    marginTop: 0,
    marginBottom: rem(12),
    fontSize: rem(14),
    lineHeight: rem(20),
    color: COLORS.textWhite,
  },
  infoGroup: { flex: 1 },
  avatarWrapper: {
    flex: "none",
    backgroundColor: COLORS.bgWhite,
    overflow: "hidden",
    borderRadius: "50%",
    width: rem(50),
    height: rem(50),
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    marginRight: rem(12),
  },
  avatar: { width: "100%" },
  title: {
    margin: [rem(16), 0, rem(8)],
    fontSize: rem(20),
    lineHeight: rem(24),
    color: COLORS.textWhite,
  },
  listChannels: {
    margin: 0,
    padding: 0,
    listStyle: "none",
  },
  channelItem: {},
  link: {
    display: "block",
    color: COLORS.textGray,
    textDecoration: "none",
    padding: [rem(12)],
    fontSize: rem(16),
    lineHeight: rem(24),
    "&.active": {
      fontWeight: "bold",
      color: COLORS.textWhite,
    },
  },

  [`@media only screen and (max-width: ${BREAKPOINTS.maxLg}px)`]: {
    root: { width: rem(260) },
  },
  [`@media only screen and (max-width: ${BREAKPOINTS.maxMd}px)`]: {
    root: { width: rem(240), padding: [rem(16), rem(8), rem(24)] },
    infoWrapper: { marginBottom: rem(24) },
    infoTitle: { fontSize: rem(12), lineHeight: rem(16) },
    avatarWrapper: { width: rem(40), height: rem(40), marginRight: rem(8) },
    title: { fontSize: rem(16), lineHeight: rem(20) },
    link: { padding: [rem(8)], fontSize: rem(12), lineHeight: rem(16) },
  },
  [`@media only screen and (max-width: ${BREAKPOINTS.maxSm}px)`]: {
    root: {
      position: "absolute",
      top: 0,
      left: 0,
      "&:not(.show)": { transform: "translateX(-100%)" },
    },
    toggleButton: {
      display: "block",
      position: "absolute",
      right: 0,
      width: rem(50),
      height: rem(50),
      transform: "translateX(100%)",
      color: COLORS.textWhite,
      cursor: "pointer",
      backgroundColor: COLORS.bgBlack,
      border: "none",
    },
  },
});
