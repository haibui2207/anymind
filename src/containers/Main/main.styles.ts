import { createUseStyles } from "react-jss";

import { rem } from "../../utils/jss";
import COLORS from "../../constants/COLORS";
import BREAKPOINTS from "../../constants/BREAKPOINTS";

export default createUseStyles({
  root: {
    display: "flex",
    boxSizing: "border-box",
    flexDirection: "column",
    height: "100vh",
    overflow: "hidden",
  },
  header: {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    padding: rem(16),
    flex: "none",
    height: rem(40),
    backgroundColor: COLORS.bgBlack01,
  },
  logo: { display: "block", height: "100%", flex: "none" },
  toggleButton: { display: "none" },
  content: {
    flex: 1,
    display: "flex",
    overflow: "hidden",
    position: "relative",
  },

  [`@media only screen and (max-width: ${BREAKPOINTS.maxSm}px)`]: {
    header: { justifyContent: "space-between" },
    logo: { height: "80%" },
    toggleButton: {
      display: "flex",
      marginLeft: rem(16),
      backgroundColor: COLORS.bgWhite,
      border: "none",
      borderRadius: rem(4),
      width: rem(30),
      height: rem(30),
      alignItems: "center",
      justifyContent: "center",
      cursor: "pointer",
    },
  },
});
