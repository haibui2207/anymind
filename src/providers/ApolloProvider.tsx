import React from 'react';
import { ApolloClient, InMemoryCache, ApolloProvider } from '@apollo/client';

const client = new ApolloClient({
  uri: 'https://angular-test-backend-yc4c5cvnnq-an.a.run.app/graphql',
  cache: new InMemoryCache(),
});

interface IProps {
  children: React.ReactNode;
}

const Provider: React.FC<IProps> = ({ children }) => (
  <ApolloProvider client={client}>{children}</ApolloProvider>
);

export default Provider;
