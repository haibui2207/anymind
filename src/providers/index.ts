export { default as ApolloProvider } from "./ApolloProvider";
export { default as ConversationProvider } from "./ConversationProvider";
