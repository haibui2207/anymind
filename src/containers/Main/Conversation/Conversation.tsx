import React, { memo, useCallback, useEffect } from "react";
import { useMutation } from "@apollo/client";
import classNames from "classnames";
import { CKEditor } from "@ckeditor/ckeditor5-react";
import ClassicEditor from "@ckeditor/ckeditor5-build-classic";

import { apiUtil } from "../../../utils";
import { useConversationContext } from "../../../providers/ConversationProvider";

import Empty from "./Empty";
import Messages from "./Messages";
import useStyles from "./conversation.styles";

const Conversation: React.FC<{}> = () => {
  const classes = useStyles();
  const {
    messages,
    draftMessage,
    activeUser,
    activeChannel,
    setDraftMessage,
  } = useConversationContext();
  

  const [postMessage, { called, loading, error }] = useMutation(
    apiUtil.POST_MESSAGE
  );

  useEffect((): void => {
    // TODO handle error if exist
    // TODO reset draft message after sending
  }, [called, loading, error]);

  // TODO using debounce if necessary
  const handleEditorChange = useCallback(
    (text: string): void => {
      setDraftMessage({
        channelId: activeChannel?.id || "",
        userId: activeUser?.id || "",
        text,
      });
    },
    [activeUser, activeChannel]
  );

  const handleSubmit = useCallback((): void => {
    if (!loading && draftMessage) {
      postMessage({
        variables: {
          channelId: activeChannel?.id || "",
          userId: activeUser?.id || "",
          text: draftMessage,
        },
      });
    }
  }, [loading, activeUser, activeChannel, draftMessage]);

  return (
    <div className={classes.root}>
      <h2 className={classes.header} data-test="conversation-channel-name">
        {activeChannel?.name}
      </h2>
      <div className={classNames("custom-scrollbar", classes.content)}>
        {messages.length > 0 ? <Messages /> : <Empty />}
      </div>
      <div className={classes.chatBox}>
        <CKEditor
          editor={ClassicEditor}
          data={draftMessage?.text || ""}
          onChange={(event, editor) => {
            const data = editor.getData();
            handleEditorChange(data);
          }}
          config={{
            placeholder: "Type your message here...",
          }}
        />
        <button
          className={classes.sendButton}
          title="Send message"
          disabled={loading || !draftMessage?.text}
          onClick={handleSubmit}
          data-test="conversation-submit-button"
        >
          <i className="icon-plane" />
        </button>
      </div>
    </div>
  );
};

export default memo(Conversation);
