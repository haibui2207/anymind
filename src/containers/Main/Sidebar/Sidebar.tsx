import React, { memo, useCallback, useEffect, useRef } from "react";
import classNames from "classnames";

import { Select } from "../../../components";
import { IChannel, IUser } from "../../../models";
import { useConversationContext } from "../../../providers/ConversationProvider";

import useStyles from "./sidebar.styles";

const Sidebar: React.FC<{}> = () => {
  const ref = useRef<HTMLDivElement>(null);
  const classes = useStyles();
  const {
    users,
    channels,
    activeUser,
    activeChannel,
    activeSidebar,
    setActiveUser,
    setActiveChannel,
    setActiveSidebar,
  } = useConversationContext();

  const handleChannelClick = useCallback(
    (channel: IChannel): void => {
      if (activeChannel?.id !== channel.id) {
        setActiveChannel(channel);
        setActiveSidebar(false);
        // TODO fetch latest message when channel change
      }
    },
    [activeChannel]
  );

  const handleUserSelectChange = useCallback(
    (user: IUser): void => {
      if (activeUser?.id !== user.id) {
        setActiveUser(user);
        // TODO fetch latest message when user change
      }
    },
    [activeUser]
  );

  useEffect(() => {
    const listener = (event: any): any => {
      if (!ref.current || ref.current.contains(event.target)) {
        return;
      }

      if (activeSidebar) {
        setActiveSidebar(!activeSidebar);
      }
    };

    document.addEventListener("mousedown", listener);
    document.addEventListener("touchstart", listener);

    return () => {
      document.removeEventListener("mousedown", listener);
      document.removeEventListener("touchstart", listener);
    };
  }, [ref, activeSidebar]);

  return (
    <div
      ref={ref}
      className={classNames("custom-scrollbar", classes.root, {
        show: activeSidebar,
      })}
    >
      <div className={classes.infoWrapper}>
        <div className={classes.avatarWrapper}>
          <img
            alt={activeUser?.name}
            src={activeUser?.avatar}
            className={classes.avatar}
            data-test="sidebar-active-user-avatar"
          />
        </div>
        <div className={classes.infoGroup}>
          <h4 className={classes.infoTitle}>Switch user</h4>
          <Select
            name="user"
            options={users}
            onChange={({ value }: { value: IUser }) =>
              handleUserSelectChange(value)
            }
            value={activeUser}
            getOptionLabel={(user: IUser): string => user.name}
            getOptionValue={(user: IUser): string => user.name}
          />
        </div>
      </div>
      <h4 className={classes.title}>Channels</h4>
      <ul className={classes.listChannels}>
        {channels.map((channel: IChannel) => (
          <li
            key={channel.id}
            className={classes.channelItem}
            data-test="sidebar-channel-item"
          >
            <a
              href="/#"
              type="button"
              className={classNames(classes.link, {
                active: activeChannel?.id === channel.id,
              })}
              onClick={(e) => {
                e.preventDefault();
                handleChannelClick(channel);
              }}
            >
              #&nbsp;&nbsp;{channel.name}
            </a>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default memo(Sidebar);
