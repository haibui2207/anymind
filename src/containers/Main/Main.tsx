import React, { memo, useCallback, useEffect } from "react";
import { useLazyQuery } from "@apollo/client";

import { apiUtil } from "../../utils";
import { IChannel, IUser, IMessage } from "../../models";
import { useConversationContext } from "../../providers/ConversationProvider";
import avatar01 from "../../assets/images/user-01.svg";
import avatar02 from "../../assets/images/user-02.svg";
import avatar03 from "../../assets/images/user-03.svg";

import useStyles from "./main.styles";
import Sidebar from "./Sidebar";
import Conversation from "./Conversation";

const Main: React.FC<{}> = () => {
  const classes = useStyles();
  const {
    users,
    channels,
    messages,
    activeUser,
    activeChannel,
    activeSidebar,
    setUsers,
    setChannels,
    setMessages,
    setActiveUser,
    setActiveChannel,
    setActiveSidebar,
  } = useConversationContext();

  const [fetchLatestMessages, { called, loading, error, data }] = useLazyQuery<
    IMessage[]
  >(apiUtil.FETCH_LATEST_MESSAGES);
  // const [fetchMoreMessages, { called, loading, error, data }] = useLazyQuery<
  //   IMessage[]
  // >(apiUtil.FETCH_MORE_MESSAGES);

  const initConversationValue = useCallback((): void => {
    const mockUsers: IUser[] = [
      { id: "Sam", name: "Sam", avatar: avatar01 },
      { id: "Russell", name: "Russell", avatar: avatar02 },
      { id: "Joyse", name: "Joyse", avatar: avatar03 },
    ];
    const mockChannels: IChannel[] = [
      { id: "1", name: "General Channel" },
      { id: "2", name: "Technology Channel" },
      { id: "3", name: "LGTM Channel" },
    ];
    // TODO Messages will be fetch from api
    const mockMessages: IMessage[] = [
      {
        messageId: "8893628174113670197",
        text: "Text from Russell",
        datetime: "2021-09-29T03:51:48.888Z",
        userId: "Russell",
      },
      {
        messageId: "4640626115768591696",
        text: "Text from Joyse",
        datetime: "2021-09-29T03:51:43.74Z",
        userId: "Joyse",
      },
      {
        messageId: "7619097256175680316",
        text: "Text from Sam",
        datetime: "2021-09-29T03:51:06.783Z",
        userId: "Sam",
      },
      {
        messageId: "8114703596647458087",
        text: "Sup",
        datetime: "2021-09-29T01:39:08.665Z",
        userId: "Sam",
      },
      {
        messageId: "459809310856795524",
        text: "Hey",
        datetime: "2021-09-29T01:02:40.527Z",
        userId: "Russell",
      },
      {
        messageId: "690924059307461292",
        text: "Hello",
        datetime: "2021-09-29T01:01:29.092Z",
        userId: "Russell",
      },
      {
        messageId: "459809310856795523",
        text: "Hey",
        datetime: "2021-09-29T01:02:40.527Z",
        userId: "Russell",
      },
      {
        messageId: "690924059307461295",
        text: "Hello",
        datetime: "2021-09-29T01:01:29.092Z",
        userId: "Russell",
      },
      {
        messageId: "459809310856795527",
        text: "Hey",
        datetime: "2021-09-29T01:02:40.527Z",
        userId: "Russell",
      },
      {
        messageId: "690924059307461492",
        text: "Hello",
        datetime: "2021-09-29T01:01:29.092Z",
        userId: "Russell",
      },
    ];
    if (!users.length) setUsers(mockUsers);
    if (!channels.length) setChannels(mockChannels);
    if (!activeUser) setActiveUser(mockUsers[0]);
    if (!activeChannel) {
      setActiveChannel(mockChannels[0]);
      if (!messages.length) {
        // TODO fetch message from api
        // fetchLatestMessages({ variables: { channelId: mockChannels[0] } });
        setMessages(mockMessages);
      }
    } else {
      if (!messages.length) {
        // TODO fetch message from api with existed channel
        // fetchLatestMessages({ variables: { channelId: activeChannel.id } });
        setMessages(mockMessages);
      }
    }
  }, [users, channels, messages, activeUser, activeChannel]);

  useEffect((): void => {
    initConversationValue();
  }, [initConversationValue]);

  useEffect((): void => {
    if (called && !loading) {
      if (!error && data) {
        setMessages(data);
      } else {
        // TODO show error cannot get list messages
        window.alert("Cannot fetch messages");
        console.error(error);
      }
    }
  }, [called, loading, error, data]);

  useEffect((): void => {
    setInterval((): void => {
      // TODO fetch more message every 15s
    }, 15000);
  }, [activeChannel]);

  const handleToggleMenu = useCallback((): void => {
    setActiveSidebar(!activeSidebar);
  }, [activeSidebar]);

  return (
    <div className={classes.root}>
      <div className={classes.header} data-test="main-header">
        <img
          alt="Anymind"
          src="https://anymindgroup.com/wp-content/themes/anymind/img/logo_white.svg"
          className={classes.logo}
        />
        {!activeSidebar ? (
          <button className={classes.toggleButton} onClick={handleToggleMenu}>
            <i className="icon-hamberger" />
          </button>
        ) : null}
      </div>
      <div className={classes.content}>
        <Sidebar />
        <Conversation />
      </div>
    </div>
  );
};

export default memo(Main);
