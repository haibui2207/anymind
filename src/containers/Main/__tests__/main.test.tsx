import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import {
  IConversation,
  ConversationContext,
  defaultContextValue,
} from "../../../providers/ConversationProvider";
import { ApolloProvider } from "../../../providers";
import Main from "../Main";
import Sidebar from "../Sidebar";
import Conversation from "../Conversation";

describe("Testing Main", (): void => {
  let wrapper: any;
  const mockContextValue: IConversation = {
    ...defaultContextValue,
    users: [{ id: "1", name: "User 1", avatar: "" }],
    channels: [{ id: "1", name: "Channel 1" }],
    messages: [
      {
        messageId: "1",
        text: "The first message",
        datetime: "2021-09-29T01:39:08.665Z",
        userId: "Sam",
      },
    ],
  };

  beforeAll((): void => {
    wrapper = mount(
      <ConversationContext.Provider value={mockContextValue}>
        <ApolloProvider>
          <Main />
        </ApolloProvider>
      </ConversationContext.Provider>
    );
  });

  it("should render match snapshot", (): void => {
    const wrapper = shallow(
      <ConversationContext.Provider value={mockContextValue}>
        <ApolloProvider>
          <Main />
        </ApolloProvider>
      </ConversationContext.Provider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should have a header", (): void => {
    expect(wrapper.find('[data-test="main-header"]')).toHaveLength(1);
  });

  it("should have a sidebar section", (): void => {
    expect(wrapper.find(Sidebar)).toHaveLength(1);
  });

  it("should have a Conversation section", (): void => {
    expect(wrapper.find(Conversation)).toHaveLength(1);
  });
});
