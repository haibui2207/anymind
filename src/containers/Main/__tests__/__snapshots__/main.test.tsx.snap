// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Main should render match snapshot 1`] = `
<Provider>
  <Memo(Main) />
</Provider>
`;
